### DISCUSSION TITLE
How can I draw a list of items at specific positions on a canvas?

### DISCUSSION THREAD

**Discussion by DrDryg** [AvaloniaUI/Avalonia#12999] [General]
I have a list of objects I like to place on a canvas based on a stored position of the objects. Basically I want to do what was suggested in the stack overflow thread: https://stackoverflow.com/questions/76005932/binding-a-list-of-rectangles-to-a-canvas-in-avaloniaui

```
    <ItemsControl Name="itemsControl" ItemsSource="{Binding Nodes}">
        <ItemsControl.ItemsPanel>
            <ItemsPanelTemplate>
                <Canvas />
            </ItemsPanelTemplate>
        </ItemsControl.ItemsPanel>

        <ItemsControl.Styles>
            <Style Selector="ContentPresenter">
                <Setter Property="Canvas.Left" Value="{Binding Node.Position.X}" />
                <Setter Property="Canvas.Top" Value="{Binding Node.Position.Y}" />
            </Style>
        </ItemsControl.Styles>

        <ItemsControl.ItemTemplate>
            <DataTemplate>
                <v:NodeView  />
            </DataTemplate>
        </ItemsControl.ItemTemplate>
    </ItemsControl>
```
My issue is that inside the style I bind to the list of nodes, not an individual node (that contains its position) in the list. Neither can I set the position in the DataTemplate since it is not a direct child of the Canvas. So Im kind of lost, how can I place my nodes correctly?

**Comment by ahopper** [comment]
see this https://github.com/AvaloniaUI/Avalonia.Samples/blob/42ab92a65d12225f1198767204139efbb38799f1/src/Avalonia.Samples/Drawing/BattleCity/MainWindow.axaml#L10 to bind the position of the ContentPresenter

**Reply by DrDryg** [reply]
Thanks, it solved my problem. I also managed to solve it using the ItemContainerTheme:

```
    <ItemsControl ItemsSource="{Binding Nodes}">

        <ItemsControl.ItemsPanel>
            <ItemsPanelTemplate>
                <Canvas />
            </ItemsPanelTemplate>
        </ItemsControl.ItemsPanel>
		
        <ItemsControl.ItemContainerTheme>
            <ControlTheme x:DataType="vm:NodeViewModel" TargetType="ContentPresenter">
                <Setter Property="Canvas.Left" Value="{Binding Node.Position.X}" />
                <Setter Property="Canvas.Top" Value="{Binding Node.Position.Y}" />
            </ControlTheme>
        </ItemsControl.ItemContainerTheme>

        <ItemsControl.ItemTemplate>
            <DataTemplate DataType="vm:NodeViewModel">
                <v:NodeView  />
            </DataTemplate>
        </ItemsControl.ItemTemplate>

    </ItemsControl>

```

One thing Im unsure of: Why cant I place the ItemsControl in a canvas and skip:

```
        <ItemsControl.ItemsPanel>
            <ItemsPanelTemplate>
                <Canvas />
            </ItemsPanelTemplate>
        </ItemsControl.ItemsPanel>
```

If Im not to place my controls at specific positions, just putting the Itemscontrol in a Canvas also seems to work.
